network/p2p/subscription_filter.go

```diff
@@ -0,0 +1,79 @@
+package p2p
+
+import (
+	"strings"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/id"
+	"github.com/onflow/flow-go/network"
+)
+
+type Filter struct {
+	idProvider  id.IdentityProvider
+	myPeerID    peer.ID
+	rootBlockID flow.Identifier
+	chainID     flow.ChainID
+}
+
+func NewSubscriptionFilter(pid peer.ID, rootBlockID flow.Identifier, chainID flow.ChainID, idProvider id.IdentityProvider) *Filter {
+	return &Filter{
+		idProvider,
+		pid,
+		rootBlockID,
+		chainID,
+	}
+}
+
+func (f *Filter) allowedTopics(pid peer.ID) map[network.Topic]struct{} {
+	id, found := f.idProvider.ByPeerID(pid)
+	var channels network.ChannelList
+
+	if !found {
+		channels = engine.UnstakedChannels()
+		// TODO: eventually we should have block proposals relayed on a separate
+		// channel on the public network. For now, we need to make sure that
+		// unstaked nodes can subscribe to the block proposal channel.
+		channels = append(channels, engine.ReceiveBlocks)
+	} else {
+		channels = engine.ChannelsByRole(id.Role)
+	}
+
+	topics := make(map[network.Topic]struct{})
+
+	for _, ch := range channels {
+		consensusCluster := engine.ChannelConsensusCluster(f.chainID)
+		syncCluster := engine.ChannelSyncCluster(f.chainID)
+
+		if strings.HasPrefix(consensusCluster.String(), ch.String()) {
+			ch = consensusCluster
+		} else if strings.HasPrefix(syncCluster.String(), ch.String()) {
+			ch = syncCluster
+		}
```

Comment:
If I understand `ChannelConsensusCluster` and `ChannelSyncCluster` correctly, their output will _always_ be prefixed, respectively, by `consensusClusterPrefix` and `syncClusterPrefix`. And granted, these aren't public constants. 

But here's what I don't get:
- if the collection node here is subscribed to the consensus cluster channel, why isn't the channel returned by `ChannelsByRole` ok to allow? Wouldn't it itself have been obtained dynamically through `ChannelConsensusCluster` at subscription, and if so, why replay this here?
- same Q for the `ChannelSyncConsensusCluster`.
- would we save time here by breaking out of the loop if this is not a collection node?

Reply:
For collection nodes, the `ChannelsByRole` function only returns the cluster prefixes, not the actual channel (which also contains the chain ID).

What we want to return from `allowedTopics` is the *exact* list of topics that can be subscribed. In particular, we want the full cluster channels, not just the prefix.

The logic here basically just replaces the cluster prefix returned from `ChannelsByRole` with the actual full channel name.

We don't want to break out of the loop since we still want to iterate through all channels and put the topic into the `topics` map.

Reply:
Thanks for these precisions!

> We don't want to break out of the loop since we still want to iterate through all channels and put the topic into the topics map.
 
Fair enough. My point is that we're still doing :
1.  constructing the full `Channel(Sync|Consensus)Cluster` name,
2. a string prefix search,
on each iteration. While:
- if I read `ClusterChannel` correctly, there's a static `(consensus|sync)ClusterPrefix` that would make the matching part cheaper, and the full channel name construction as-needed.
- the cluster-matching logic only seems necessary if the role of the node relies on clusters [1], 

[1]: Maybe `channelRoleMap[(consensus|sync)Cluster]Prefix` could be compared against the node role?

Reply:
> if I read ClusterChannel correctly, there's a static (consensus|sync)ClusterPrefix that would make the matching part cheaper, and the full channel name construction as-needed.

Yeah, but those are private variables and therefore not able to be directly compared against.

But anyway, I think overall you're right and I can make some changes to this logic here.